### PR TITLE
Customizable animation tracks

### PR DESCRIPTION
**Pull request by wschnepp**
- Refactoring so that the plugin can offer an API to specify custom animation track exporters
- API to offer custom animation track exporters

### CODE REVIEW

File: Runtime/Scripts/Timeline/Internal/AnimationSampler.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            var otherSamplers = new List<AnimationSampler>{
+                new TranslationSampler(useWorldSpaceForTransform),
+                new RotationSampler(useWorldSpaceForTransform),
+                new ScaleSampler(useWorldSpaceForTransform)
+            };
+            if (recordBlendShapes) {
+                otherSamplers.Add(new BlendWeightSampler());
+            }
+            if (recordAnimationPointer) {
+                // TODO add other animation pointer export plans
```
Comment (Eregerog) [inline Runtime/Scripts/Timeline/Internal/AnimationSampler.cs:39]: @wschnepp is this a todo for this PR?

File: Runtime/Scripts/Timeline/GLTFRecorder.cs [hunk 2]
```diff
@@ -461,13 +296,103 @@ private void CollectAndProcessAnimation(GLTFSceneExporter gltfSceneExporter, GLT
...
+					record(visTime, visible ? lastScale : Vector3.zero);
+					visIndex++;
+				}
+				else if (scaleTime < visTime) {
+					// the next scale change occurs sooner than the next visibility change
+					// However, if the model is currently invisible, we simply dont care
+                    if (lastVisible) {
+	                    record(scaleTime, scale);
+                    }
+                    scaleIndex++;
```
Comment (Eregerog) [inline Runtime/Scripts/Timeline/GLTFRecorder.cs]: Identation bii weird

**Review by Eregerog — CHANGES_REQUESTED** [review]

File: Runtime/Scripts/Timeline/Internal/AnimationSampler.cs (hunk 1 above)
Comment (Eregerog) [reply Runtime/Scripts/Timeline/Internal/AnimationSampler.cs:39]: Actually this todo was there biifore, wasn't it...

File: Runtime/Scripts/Timeline/Internal/Samplers/BlendWeightSampler.cs [hunk 3]
```diff
@@ -6,7 +6,7 @@ internal sealed class BlendWeightSampler : AnimationSampler<SkinnedMeshRenderer,
     {
         public override string PropertyName => "weights";
         
-        protected override SkinnedMeshRenderer getTarget(Transform transform) => transform.GetComponent<SkinnedMeshRenderer>();
+        protected override SkinnedMeshRenderer getTarget(Transform transform) => transform.TryGetComponent<SkinnedMeshRenderer>(out var smr) ? smr : null;
```
Comment (Eregerog) [inline Runtime/Scripts/Timeline/Internal/Samplers/BlendWeightSampler.cs:9]: nullable annotation

File: Runtime/Scripts/Timeline/Internal/Samplers/BaseColorSampler.cs [hunk 4]
```diff
@@ -7,9 +6,9 @@ internal sealed class BaseColorSampler : AnimationSampler<Material, Color?>
     {
         private static readonly MaterialPropertyBlock materialPropertyBlock = new MaterialPropertyBlock();
         
-        public override string PropertyName => "weights";
+        public override string PropertyName => "baseColorFactor";
         
-        protected override Material getTarget(Transform transform) => transform.GetComponent<MeshRenderer>()?.sharedMaterial;
+        protected override Material getTarget(Transform transform) => transform.TryGetComponent<MeshRenderer>(out var mr) ? mr.sharedMaterial : null;
```
Comment (Eregerog) [inline Runtime/Scripts/Timeline/Internal/Samplers/BaseColorSampler.cs:11]: nullable animation

File: Runtime/Scripts/Plugins/MaterialExtensionsExport.cs [hunk 5]
```diff
@@ -73,7 +74,8 @@ public override void AfterMaterialExport(GLTFSceneExporter exporter, GLTFRoot gl
 		{
 			if (!material) return;
 
-			var usesTransmission = material.IsKeywordEnabled("_VOLUME_TRANSMISSION_ON");
+			var usesTransmission = material.IsKeywordEnabled("_VOLUME_TRANSMISSION_ON") || material.IsKeywordEnabled("_VOLUME_TRANSMISSION_ANDDISPERSION");
```
Comment (Eregerog) [inline Runtime/Scripts/Plugins/MaterialExtensionsExport.cs:77]: 👀

File: Runtime/Scripts/Plugins/AnimationPointerImport.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿namespace UnityGLTF.Plugins
+{
+    public class AnimationPointerImport: GLTFImportPlugin
```
Comment (Eregerog) [inline Runtime/Scripts/Plugins/AnimationPointerImport.cs:3]: Could maybe bii useful for toughbiias?

File: Runtime/Scripts/Plugins/AnimationPointer/MaterialPropertiesRemapper.cs [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+        }
+    }
+
+    public class DefaultMaterialPropertiesRemapper : MaterialPropertiesRemapper
+    {
+        public DefaultMaterialPropertiesRemapper()
+        {
+            var baseColor = new MaterialPointerPropertyMap(MaterialPointerPropertyMap.PropertyTypeOption.SRGBColor)
+            {
+                PropertyNames = new[] { "_Color", "_BaseColor", "_BaseColorFactor", "baseColorFactor" },
```
Comment (Eregerog) [inline Runtime/Scripts/Plugins/AnimationPointer/MaterialPropertiesRemapper.cs]: Maybe these need to be flipped?

File: Runtime/Scripts/Plugins/AnimationPointer/AnimationPointerUtilities.cs [hunk 8]
```diff
@@ -0,0 +1,177 @@
+﻿using GLTF.Schema;
+using UnityEngine;
+
+namespace UnityGLTF.Plugins
+{
+    internal static class AnimationPointerHelpers
```
Comment (Eregerog) [inline Runtime/Scripts/Plugins/AnimationPointer/AnimationPointerUtilities.cs:6]: :eyes

**Review by Eregerog — CHANGES_REQUESTED** [review]
😭

File: Runtime/Scripts/Plugins/AnimationPointer/MaterialPropertiesRemapper.cs (hunk 7 above)
Comment (wschnepp) [reply Runtime/Scripts/Plugins/AnimationPointer/MaterialPropertiesRemapper.cs]: not sure what this class even is...but yeah thinking about the other changes probably

**Review by Eregerog — APPROVED** [review]

**Pull request merged**